web/docs/pipelines.md

```diff
@@ -0,0 +1,216 @@
+---
+sidebar_position: 0
+---
+
+# Pipelines
+
+A Tenzir **pipeline** is a chain of **operators** that represents a dataflow.
+Operators are the atomic building blocks that produce, transform, or consume
+data. Think of them as UNIX or Powershell commands where output from one command
+is input to the next:
+
+![Pipeline Chaining](pipeline-chaining.excalidraw.svg)
+
+There exist three types of operators: **sources** that produce data, **sinks**
+that consume data, and transformations that do both:
+
+![Pipeline Structure](pipeline-structure.excalidraw.svg)
+
+## Polymorphic Operators
+
+Tenzir pipelines make one more distinction: the elements that the operators push
+through the pipeline are *typed*. Every operator has an input and an output
+type:
+
+![Input and Output Types](operator-pieces.excalidraw.svg)
+
+When composing pipelines out of operators, the type of adjacent operators have
+to match. Otherwise the pipeline is malformed. Here's an example pipeline with
+matching operators:
+
+![Typed Pipeline](typed-pipeline.excalidraw.svg)
+
+We call any void-to-void operator sequence a **closed pipeline**. Only closed
+pipelines can execute. If a pipeline does not have a source and sink, it would
+"leak" data. If a pipeline is open, the engine auto-completes a source/sink when
+possible or rejects it.
```

Comment:
We should consider mentioning here that the default source/sink depends on how you use Tenzir. In the app, we automatically show the results in the explorer if the sink is missing. On the command-line, we read/write JSON from/to stdin.

Reply:
Changed to:

> If a pipeline is open, the engine auto-completes a source/sink when
possible and rejects the pipeline otherwise. Auto-completion is context
dependent: on the command line we read JSON from stdin and write it stdout. In
the app we only auto-complete a missing sink with
[`serve`](operators/sinks/serve.md) to display the result in the browser.